### PR TITLE
Red chapter 5 exercises

### PR DESCRIPTION
**Pull request by utopia-2**
Chapter 5 exercises, answers and _Try This_ tweaks.

### CODE REVIEW

File: src/org/java/training/Average.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package org.java.training;
+
+public class Average {
+	
+	public static void main(String[] args) {
+		double[] vals = { 24.03, 25_325.34, 22.0, 1.5, -3_994.543, 2_435.33, 100.00, 293.34, 34, 2_385.34 };
```
Comment (Spinodal) [inline src/org/java/training/Average.java:6]: `_`'s ftw!!!

File: src/org/java/training/StringSort.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+import java.util.Random;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class StringSort {
+	
+	public static void main(String[] args) {
+		String[] message = { "This", "is", "an", "array", "of", "strings", "sort", "them", "please.", "I'll", "wait..." };
+		Supplier<String> join = () -> Arrays.stream(message).collect(Collectors.joining(" "));
```
Comment (Spinodal) [inline src/org/java/training/StringSort.java:16]: greatly confused...

File: answers.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+```java
+String like = "I like Java";
+int len = like.length();
+char ch = like.charAt(4);
+```
+
+Or without assigning to a reference:
+
+```java
+"I like Java".charAt(3);
```
Comment (Spinodal) [inline answers.md:45]: I think it was wanting this way to illustrate you can call a method on a string like this

File: src/org/java/training/BadEncode.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+		StringBuilder enc = new StringBuilder(), dec = new StringBuilder();
+		for(int i = 0, lim = key.length(); i < msg.length(); enc.append((char)( key.charAt(i % lim) ^ msg.charAt(i++) ) ) );
+		
+		System.out.println("Encoded message: " + enc);
+		
+		for(int i = 0, lim = key.length(); i < enc.length(); dec.append((char)( key.charAt(i % lim) ^ enc.charAt(i++) ) ) );
+		
+		System.out.println("Decoded message: " + dec);
+	}
+}
```
Comment (Spinodal) [inline src/org/java/training/BadEncode.java:20]: too many one liners!!!

File: answers.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+ 13) _Rewrite the `MinMax` class shown earlier in this chapter so that it uses a for-each style `for` loop._
+ 
+ See [MinMax.java](src/org/java/training/MinMax.java).
+ 
+ 14) _Can the `for` loops that perform sorting in the `Bubble` class shown in `Try This 5-1` be converted into
+     for-each style loops? If not, why not?_
+     
+No. Sorting algorithms operate by comparing and swapping values held at different indices in each iteration
+whereas the for-each loop operates on one value at a time.
```
Comment (Spinodal) [inline answers.md:92]: I think you could technically sort the elements into a new array/list, but that might not be the most efficient...

Comment (utopia-2) [inline answers.md:92]: You could insert it into a self-ordering structure like a binary tree, but then it's not the same as the for-each loop doing the sorting.

File: src/org/java/training/BadEncode.java (hunk 4 above)
Comment (utopia-2) [inline src/org/java/training/BadEncode.java:20]: The right amount! :sunglasses:

File: src/org/java/training/StringSort.java (hunk 2 above)
Comment (utopia-2) [inline src/org/java/training/StringSort.java:16]: It's an object that wraps the `message` array to make it easy to later print out the arrays contents with just `join.get()`. `collect` takes a `Collector` that builds up a result from each element of a stream, `Collectors.joining` provides a `Collector` that joins each `String`, placing the delimiter between each.

File: answers.md (hunk 3 above)
Comment (utopia-2) [inline answers.md:45]: I didn't think that'd be what it was after, as it didn't really get into string literals as object references in the chapter. Taking a look, this _is_ what it wanted though - ah well, adding both is always a safe bet :smile:.

File: src/org/java/training/BadEncode.java (hunk 4 above)
Comment (utopia-2) [inline src/org/java/training/BadEncode.java:20]: The duplicate behaviour could be reduced, but I wanted to more or less follow the book, with improvements:

``` java
Function<String, Supplier<String>> cipher = s -> {
  int j = 0;
  return () -> msg.chars().map(c -> (char)c ^ s.charAt(j++ % s.length())
                          .map(c -> c.toString()).collect(Collectors.joining());
  };
```

Then later:

``` java
String encoded = cipher.accept(msg).get();
String decoded = cipher.accept(encoded).get();
```

**Pull request closed without merging**